### PR TITLE
Refactor createTransactionHash and MerkleHashBuilder

### PR DESCRIPTION
**Pull request by evias**
### Changed

- Refactor of Transaction.createTransactionHash for clarity and use protocol nearest transaction creation flow as seen in [TransactionExtensions](https://github.com/nemtech/catapult-server/blob/master/sdk/src/extensions/TransactionExtensions.cpp#L46)
- MerkleHashBuilder now creates hasher from instance parameters
- Fix usage of **SHA3** for transaction hashes in `Transaction.createTransactionHash` and `AggregateTransaction.calculateInnerTransactionHash

### CODE REVIEW

File: src/model/transaction/AggregateTransaction.ts [hunk 1]
```diff
@@ -372,11 +372,23 @@ export class AggregateTransaction extends Transaction {
      * Generate inner transaction root hash (merkle tree)
      * @returns {Uint8Array}
      */
-    private calculateInnerTransactionHash(): Uint8Array {
-        const builder = new MerkleHashBuilder(SHA3Hasher.createHasher);
+    private calculateInnerTransactionHash(networkType: NetworkType): Uint8Array {
```
Comment (rg911) [inline src/model/transaction/AggregateTransaction.ts]: Seems networkType is not used at all. should it be removed?

File: src/model/transaction/Transaction.ts [hunk 2]
```diff
@@ -81,29 +111,68 @@ export abstract class Transaction {
...
+     *
+     * @see https://github.com/nemtech/catapult-server/blob/master/src/catapult/model/EntityHasher.cpp#L32
+     * @see https://github.com/nemtech/catapult-server/blob/master/src/catapult/model/EntityHasher.cpp#L35
+     * @see https://github.com/nemtech/catapult-server/blob/master/sdk/src/extensions/TransactionExtensions.cpp#L46
      * @param {string} transactionPayload HexString Payload
      * @param {Array<number>} generationHashBuffer Network generation hash byte
      * @param {NetworkType} networkType Catapult network identifier
      * @returns {string} Returns Transaction Payload hash
      */
     public static createTransactionHash(transactionPayload: string, generationHashBuffer: number[], networkType: NetworkType): string {
```
Comment (rg911) [inline src/model/transaction/Transaction.ts]: again, don't think we need networkType anymore.

File: src/model/transaction/AggregateTransaction.ts (hunk 1 above)
Comment (evias) [reply src/model/transaction/AggregateTransaction.ts]: that one can be removed

File: src/model/transaction/Transaction.ts (hunk 2 above)
Comment (evias) [reply src/model/transaction/Transaction.ts]: this one i would rather keep as it affects CosignatureTransaction down the road ; lets wait for more clarity on keccak for removing it ?

**Pull request merged**